core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -683,8 +682,11 @@ class GroupMetadataManager(brokerId: Int,
                       removedGroups.add(groupId)
                     }
 
-                  case unknownKey =>
-                    throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+                  case unknownKey: UnknownKey =>
+                    // Unknown versions may exist when a downgraded coordinator is reading records from the log.
+                    warn(s"Unknown message key with version ${unknownKey.version}" +
+                      s" while loading offsets and group metadata. Ignoring it. " +
```

Comment:
nit: I wonder if we could add `while loading offsets and group metadata from $topicPartition`. Having the partition may be useful as well.